Bouncy Objects

SurajDatta opened the pull request:
# Description
Hi, I have added the following codes:
- command for prepare 10 Bowling pins
- command for dropping many blocks
# How to test
- to activate bulk dropper type `bulkDrop (your_block_name) (drop_amount)`
- to activate bowling pin type `bowlingPrep (your_block_name)`

GooeyHub commented:
Can one of the admins please verify this patch?

SurajDatta commented:
bowling pin* 😅

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -76,18 +78,16 @@
-import java.util.Locale;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Please change your settings in IntelliJ so this doesn't happen. We require all imports to not use the `*`

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
         blockItem.send(new DropItemEvent(spawnPos));
         return "Spawned block.";
     }
+    @Command(shortDescription = "spawns many block together (value up to you prefer)", helpText = "you can select the particular block you want to spawn" +
```
I'd prefer this description to be re-worded to fit better with the other commands. Eg, correct grammar (If English isn't your first language feel free to ask someone to check it before you make the PR), similar third person type description as the other commands. Eg, no "you"

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+            if (block == null) {
+                return "";
+            }
+            BlockItemFactory blockItemFactory = new BlockItemFactory(entityManager);
```
Even though the compiler will move some of it outside the for loop, I'd like to see you move it all outside. Only the last two lines should be in the loop.

jellysnake replied:
These should still be outside the loop. You only need two lines inside the loop

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+
+
+            offset.scale(0);
+            // a is the distance betweeen the pins and coordinate difference in z vector
```
Again, like for the other command please move what doesn't need to be in the loop to outside

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+
+            float y = 6.0f; float x = 2.0f; float z =  2.0f; float a = 0.3f; float b = 0.5f;
+            // there is no guarantee that the pins will stand straight, but the range for ir to stand up right is from the vector y between (5.8 - 6.1)
+            float x1 = 0; float x2 = 0; float x3 = 0; float x4 = 0; float x5 = 0; float x6 = 0; float x7 = 0; float x8 = 0; float x9 = 0; float x10 = 0; float z1 = 0; float z2 = 0; float z3 = 0; float z4 = 0; float z5 = 0; float z6 = 0; float z7 = 0; float z8 = 0; float z9 = 0; float z10 = 0;
```
Please use informative names, in this particular case you probably should have used a a loop.

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+            x7 = x4+a; z7 = z6+b;
+            x8 = x2; z8 = z7;
+            x9 = x3; z9 = z8;
+            x10 = x6-a; z10 = z8;
```
See above, incredibly hard to read. I have literally no idea what this section does without spending a lot of time reading it.

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+            }
+            if (m == 9) {
+                spawnPos.add(x10, y, z10);
+            }
```
Please use a switch-case here instead of chaining if statements.

However this is pointing to you using loops incorrectly. This whole section could be simplified drastically

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        }
```
Rather than sleep twice, just sleep once for longer.

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +536,139 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        }
+
+
+        return "Dropped 10 Bowlingpin :)";
```
See previous comment on grammar

SurajDatta commented:
Hi, I have ade the following changes:
- Changed the import setting, so now there is no `*`
- Changes all the gramatical mistakes
- from the bulk dropper I have moved all the unneccesary code outside the loop, but in Bowling Pin I could not move the top 5 lines oustide. Because it is behaving wierdly; if I take those lines outside, the pins are not dropping in a regular manner. I agreewith you that this could be taken outside the loop but since I am facing the problem, I am unable to do so.
- I have changed the variables and given them a proper understandable name.
- I have added a switch case in place of multiple if statements.
- I have removed a sleep timer and have only one.

hope everything is fine now :)

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +541,126 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        }
+
+        // this returns the block you have spawned and the amount
+        return "Dropped " + value + " Blocks :)";
```
Please include the name of the block here

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +541,126 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        return "Dropped " + value + " Blocks :)";
+    }
+
+    @Command(shortDescription = "Drops bowling pin", helpText = " `prepare` command to be entered along with the `block name` ",
```
This description is incorrect, it doesn't drop a bowling pin. It arranges the blocks in a bowling pin format.
See the other commands for how the `helpText` section should be written as well.

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +541,126 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+                Thread.sleep(100);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
```
Is it needed to sleep?

jellysnake commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -516,14 +541,126 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        }
+
 
+        return "Dropped 10 Bowlingpin :)";
```
Please make this more informative, like in other commands.

SurajDatta commented:
Hi, I have made the following changes:

- Removed all the unused imports
- Rebuild the whole `prepare`
- Removed the sleep statements
- Changes the descriptions and the help text to be more informative
- Changed the return statements and added more information

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
         blockItem.send(new DropItemEvent(spawnPos));
         return "Spawned block.";
     }
+    @Command(shortDescription = "Drops many blocks as per value choosen by the player", helpText = "the particular block can be seleted which the player wants to drop"
```
Help text here is a bit weird and has a typo. Also doesn't touch the `value` param (doesn't need a full independent description or anything but could be something like "You can indicate which block to drop with the first parameter and how many with the second")
  
Could the `value` param be optional with a default value of maybe 100? Then the player could just use `bulkDrop dirt`. And on a similar note should `value` be capped? Have you tried to see what a reasonable upper bound would be? If 2000 blocks lags even a good system we might as well make maybe 10000 the max and return a message to the player if a higher number was requested.

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
     }
+    @Command(shortDescription = "Drops many blocks as per value choosen by the player", helpText = "the particular block can be seleted which the player wants to drop"
+                , runOnServer = true, requiredPermission = PermissionManager.CHEAT_PERMISSION)
+    public String bulkDropper(@Sender EntityRef sender, @CommandParam("blockName") String blockName, @CommandParam("value") int value) {
```
Maybe a better name: `bulkDrop`

Seems more natural to `bulkDrop dirt 250` than `bulkDropper dirt 250` and it is slightly shorter

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        spawnPos.add(5,10,0);
+        BlockFamily block = blockManager.getBlockFamily(blockName);
+        if (block == null) {
+            return "";
```
In case of an error please return a useful message to the user, that'll show up in the console to indicate something went wrong, hopefully with some details :-)

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+    @Command(shortDescription = "Arranges the blocks in a bowling pin format.", helpText = "Spawns a specific blocks in a regular bowling pin pattern, The item " +
+            " front of the player can simply picked up",
+            runOnServer = true, requiredPermission = PermissionManager.CHEAT_PERMISSION)
+    public String prepare  (@Sender EntityRef sender, @CommandParam("blockName") String blockName) {
```
Requesting a command rename here: `prepare` is very generic and suggests nothing about what is being prepared. Maybe instead "bowlingPrep` or `bowlingPins` or something in that vein. Look for a command that's unique enough so auto-complete will have no trouble with ambiguous leading letters. 

Text could also be improved a little bit ("Arranges the blocks" is bad grammar - which blocks? Could as well be "Sets up a typical bowling pin arrangement in front of the player" or so)

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+
+        BlockFamily block = blockManager.getBlockFamily(blockName);
+        if (block == null) {
+            return "";
```
Like the other command it would be nice to tell the user something helpful if there is an error :-)

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -506,24 +526,91 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+
+        Vector3f startPos = new Vector3f(spawnPos);
+
+        float deltax = .2f;
```
A quick comment describing these three might be handy, maybe also making the `y` variable name more descriptive (isn't that actually triggering a Checkstyle warning?)

SurajDatta commented:
made*

SurajDatta commented:
Hi, I have made the following changes:
- changed the help texts and descriptions
- caped to 5000 blocks
- made the commands shorter 
- improved the return statements
-  I have also check with the checkStyle and everything looks fine :-)
- I have tried to implement the default command, which was setting `value` to 100 when just `bulkDrop stone` is called, but it seems it is very hard to do so.

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -517,13 +537,87 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        spawnPos.add(5,10,0);
+        BlockFamily block = blockManager.getBlockFamily(blockName);
+        if (block == null)
+        {
```
Are you sure you've got Checkstyle set up within IntelliJ? Braces like this on the next line is one of the very common violations it catches. Starting braces should be on the same line as the statement that begins the new block :)

Also, you can make the return more informative by including the actual block name in the message

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -517,13 +537,87 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        }
+
+        BlockItemFactory blockItemFactory = new BlockItemFactory(entityManager);
+        if(value > 5000)
```
More Checkstyle: Should have caught missing space between `if(`

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -517,13 +537,87 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        BlockItemFactory blockItemFactory = new BlockItemFactory(entityManager);
+        if(value > 5000)
+        {
+            return "sorry the value you have entered exceeds the limit of 5000";
```
Easy to tell the user what the number was (rarely can help if a bug results in a different value than the user expected)

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -517,13 +537,87 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+    @Command(shortDescription = "Sets up a typical bowling pin arrangement in front of the player. ", helpText = "Spawns a specific blocks in a regular bowling pin pattern, The item " +
+            " front of the player can simply picked up",
+            runOnServer = true, requiredPermission = PermissionManager.CHEAT_PERMISSION)
+    public String bowlingPrep  (@Sender EntityRef sender, @CommandParam("blockName") String blockName) {
```
You still have some mystery extra spaces or lines like here between `bowlingPrep  (@...`

Cervator commented on engine/src/main/java/org/terasology/logic/console/commands/CoreCommands.java:
```diff
@@ -517,13 +537,87 @@ public String spawnBlock(@Sender EntityRef sender, @CommandParam("blockName") St
+        Vector3f startPos = new Vector3f(spawnPos);
+
+        float deltax = .2f; // delta x is the distance between the pins in vectorX coordinate
+        float deltaz = .3f; //delta z is the distance between the pins in vectorZ coordinate
```
For clarity it may help to indicate what dimension `z` is rather than just saying delta z is the dist between pins in vectorZ - that somewhat just rephrases things :)

SurajDatta commented:
Hi, I have made the following changes:
- corrected all the checkStyle issues
- rephrased some explanations
- add the value typed by the user
- Also in the return statement, I am not sure what it does hence I cant write any info on that until I understand when does that return execute

Cervator commented:
Finally had a chance to get back to this. Works pretty well, good work!

At the moment the bowling pins don't manage to stay standing, and the command seems to place stuff off to the side. But physics will change soon so maybe we can improve this some more later. I'd also be curious to be able to vary the spread on the bulk drop one day :-)

Thanks @SurajDatta !

I'm merging this via web editor with squash since the commits are a bit messy, then pushing another commit myself with slight tweaks to the help text :-)
